nifi-nar-bundles/nifi-standard-services/nifi-distributed-cache-services-bundle/nifi-distributed-cache-client-service/src/main/java/org/apache/nifi/distributed/cache/client/DistributedMapCacheClientService.java

```diff
@@ -99,411 +101,140 @@
-            dos.writeUTF("fetch");
-
-            serialize(key, keySerializer, dos);
-            dos.flush();
-
-            // read response
-            final DataInputStream dis = new DataInputStream(session.getInputStream());
-            final long revision = dis.readLong();
-            final byte[] responseBuffer = readLengthDelimitedResponse(dis);
-
-            if (revision < 0) {
-                // This indicates that key was not found.
-                return null;
-            }
-
-            return new AtomicCacheEntry(key, valueDeserializer.deserialize(responseBuffer), revision);
-        });
+    public <K, V> AtomicCacheEntry<K, V, Long> fetch(final K key, final Serializer<K> keySerializer,
+                                                     final Deserializer<V> valueDeserializer) throws IOException {
+        validateProtocolVersion(ProtocolVersion.V2.value());
+        final byte[] bytesKey = CacheClientSerde.serialize(key, keySerializer);
+        final AtomicCacheEntryInboundAdapter<K, V> inboundAdapter =
+                new AtomicCacheEntryInboundAdapter<>(key, valueDeserializer);
+        return cacheClient.fetch(bytesKey, inboundAdapter);
     }
 
-    private void validateProtocolVersion(final CommsSession session, final int requiredProtocolVersion) {
-        if (session.getProtocolVersion() < requiredProtocolVersion) {
+    private void validateProtocolVersion(final int requiredProtocolVersion) {
+        if (versionNegotiator.getVersion() < requiredProtocolVersion) {
             throw new UnsupportedOperationException("Remote cache server doesn't support protocol version " + requiredProtocolVersion);
         }
     }
 
     @Override
     public <K, V> boolean replace(AtomicCacheEntry<K, V, Long> entry, Serializer<K> keySerializer, Serializer<V> valueSerializer) throws IOException {
-        return withCommsSession(session -> {
-            validateProtocolVersion(session, 2);
-
-            final DataOutputStream dos = new DataOutputStream(session.getOutputStream());
-            dos.writeUTF("replace");
-
-            serialize(entry.getKey(), keySerializer, dos);
-            dos.writeLong(entry.getRevision().orElse(0L));
-            serialize(entry.getValue(), valueSerializer, dos);
-
-            dos.flush();
-
-            // read response
-            final DataInputStream dis = new DataInputStream(session.getInputStream());
-            return dis.readBoolean();
-        });
+        validateProtocolVersion(ProtocolVersion.V2.value());
+        final byte[] bytesKey = CacheClientSerde.serialize(entry.getKey(), keySerializer);
+        final byte[] bytesValue = CacheClientSerde.serialize(entry.getValue(), valueSerializer);
+        final long revision = entry.getRevision().orElse(DEFAULT_CACHE_REVISION);
+        return cacheClient.replace(bytesKey, bytesValue, revision);
     }
 
+    private static final long DEFAULT_CACHE_REVISION = 0L;
```

Comment:
This declaration should be moved to the beginning of the class, before constructors and methods.